library/modules/Items.cpp

```diff
@@ -2270,6 +2270,74 @@ bool Items::markForTrade(df::item *item, df::building_tradedepotst *depot) {
     return true;
 }
 
+bool Items::canMelt(df::item *item)
+{
+    CHECK_NULL_POINTER(item);
+
+    switch(item->getType()) {
+        // these are not meltable, even if made from metal
+        case df::item_type::CORPSE:
+        case df::item_type::CORPSEPIECE:
+        case df::item_type::REMAINS:
+        case df::item_type::FISH:
+        case df::item_type::FISH_RAW:
+        case df::item_type::VERMIN:
+        case df::item_type::PET:
+        case df::item_type::FOOD:
+        case df::item_type::EGG:
+        case df::item_type::BAR:
+            return false;
+        default:
+            break;
+    }
+
+    // original check from the logistics module
+    MaterialInfo mat(item);
+    if (mat.getCraftClass() != df::craft_material_class::Metal)
+        return false;
+
+    for (auto &g : item->general_refs) {
+        switch (g->getType()) {
+            case df::general_ref_type::CONTAINS_ITEM:
+            case df::general_ref_type::UNIT_HOLDER:
+            case df::general_ref_type::CONTAINS_UNIT:
+                return false;
+            case df::general_ref_type::CONTAINED_IN_ITEM:
+            {
+                df::item *c = g->getItem();
+                for (auto &gg : c->general_refs) {
+                    if (gg->getType() == df::general_ref_type::UNIT_HOLDER)
+                        return false;
+                }
+                break;
+            }
+            default:
+                break;
+        }
+    }
```

Comment:
While refactoring, I didn't question this logic too much. But it clearly does not correspond to what DF does.
1. DF allows designating equipped items for melting (at least it displays a melting icon)
2. DF allows melting non-empty metal bins (I tested with bins containing metal bars, the bars end up in the workshop)

Reply:
I contend that while this does not accurately mirror DF logic, it is appropriate for how this function is used, which is more along the lines of "should I consider this item as a candidate for melting". I think players would be confused and annoyed if they select all for melting and get their full bins melted.

Reply:
@myk002 Unless you (or someone else) objects, I would move this check back to `logistics`, just like the `melt_masterworks` check.

Reply:
Sorry, I wrote my last message without reloading the page. How about taking a boolean flag (that defaults to true) to toggle those checks that do not strictly mirror the logic from DF? At least this would clearly document the deviations.

Reply:
That would be acceptable, though make sure the flag defaults to *false*, otherwise you'll need to write a custom shim function to insert the default value when you expose to Lua.